File: frontend/lib/models/session.dart
import 'dart:convert';

class Session {
  final String id;
  final DateTime begin;
  final DateTime end;
  final String title;
  final String description;
  final String? place;
  final String kind;
  final String? companyId;
  final List<String>? speakersIds;
  final String? videoURL;
  final SessionTickets? tickets;

  Session({
    required this.id,
    required this.begin,
    required this.end,
    required this.title,
    required this.description,
    this.place,
    required this.kind,
    this.companyId,
    this.speakersIds,
    this.videoURL,
    this.tickets,
  });

  factory Session.fromJson(Map<String, dynamic> json) {
    return Session(
      id: json['id'],
      begin: DateTime.parse(json['begin']),
      end: DateTime.parse(json['end']),
      title: json['title'],
      description: json['description'],
      place: json['place'],
      kind: json['kind'],
      companyId: json['company'],
      speakersIds: List.from(json['speaker']),
      videoURL: json['videoURL'],
      tickets: json['tickets'] == null
          ? null
          : SessionTickets.fromJson(json['tickets']),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'begin': begin.toIso8601String(),
        'end': end.toIso8601String(),
        'title': title,
        'description': description,
        'place': place,
        'kind': kind,
        'company': companyId,
        'speaker': speakersIds,
        'videoURL': videoURL,
        'tickets': tickets?.toJson(),
      };

  @override
  String toString() {
    return json.encode(this.toJson());
  }
}

class SessionPublic {
  final String? id;
  final DateTime? begin;
  final DateTime? end;
  final String? title;
  final String? description;
  final String? place;
  final String? kind;
  final String? companyPublicId;
  final List<String>? speakersPublicIds;
  final String? videoURL;
  final SessionTickets? tickets;

  SessionPublic({
    this.id,
    this.begin,
    this.end,
    this.title,
    this.description,
    this.place,
    this.kind,
    this.companyPublicId,
    this.speakersPublicIds,
    this.videoURL,
    this.tickets,
  });

  factory SessionPublic.fromJson(Map<String, dynamic> json) {
    return SessionPublic(
      id: json['id'],
      begin: DateTime(json['begin']),
      end: DateTime(json['end']),
      title: json['title'],
      description: json['description'],
      place: json['place'],
      kind: json['kind'],
      companyPublicId: json['company'],
      speakersPublicIds: json['speaker'],
      videoURL: json['videoURL'],
      tickets: SessionTickets.fromJson(json['tickets']),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'begin': begin,
        'end': end,
        'title': title,
        'description': description,
        'place': place,
        'kind': kind,
        'company': companyPublicId,
        'speaker': speakersPublicIds,
        'videoURL': videoURL,
        'tickets': tickets?.toJson(),
      };

  @override
  String toString() {
    return json.encode(this.toJson());
  }
}

class SessionTickets {
  final DateTime? start;
  final DateTime? end;
  final int? max;

  SessionTickets({
    this.start,
    this.end,
    this.max,
  });

  factory SessionTickets.fromJson(Map<String, dynamic> json) {
    return SessionTickets(
      start: DateTime.parse(json['start']),
      end: DateTime.parse(json['end']),
      max: json['max'],
    );
  }

  Map<String, dynamic> toJson() => {
        'start': start?.toUtc().toIso8601String(),
        'end': end?.toUtc().toIso8601String(),
        'max': max,
      };

  @override
  String toString() {
    return json.encode(this.toJson());
  }
}

File: backend/src/models/session.go
package models

import (
	"errors"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type SessionKind string

const (
	SessionKindTalk         SessionKind = "TALK"
	SessionKindPresentation SessionKind = "PRESENTATION"
	SessionKindWorkshop     SessionKind = "WORKSHOP"
)

func (sk *SessionKind) Parse(kind string) error {

	var newSessionKind SessionKind

	switch kind {

	case string(SessionKindTalk):
		newSessionKind = SessionKindTalk
		break

	case string(SessionKindPresentation):
		newSessionKind = SessionKindPresentation
		break

	case string(SessionKindWorkshop):
		newSessionKind = SessionKindWorkshop
		break

	default:
		return errors.New("invalid kind")

	}

	*sk = newSessionKind
	return nil
}

// SessionTickets are the session's reservation tickets' details.
// For example, if a workshop is ment to be given for 20 people, then
// there should be a maximum of 20 tickets available.
type SessionTickets struct {
	Start time.Time `json:"start"`
	End   time.Time `json:"end"`
	Max   int       `json:"max"`
}

// Session is a scheduled talk, presentation or workshop. This will be used to generate
// the event's week schedule.
type Session struct {
	// Session's ID (_id of mongodb).
	ID primitive.ObjectID `json:"id" bson:"_id"`

	Begin       time.Time `json:"begin" bson:"begin"`
	End         time.Time `json:"end" bson:"end"`
	Title       string    `json:"title" bson:"title"`
	Description string    `json:"description" bson:"description"`

  // Session Image
  Image string `json:"img,omitempty" bson:"img,omitempty"`

	// Where the session is being held. Typically a room on the venue.
	Place string `json:"place" bson:"place"`

	// Kind of session can be "TALK", "PRESENTATION" or "WORKSHOP".
	Kind SessionKind `json:"kind" bson:"kind"`

	// Company is an _id of Company (see models.Company).
	Company *primitive.ObjectID `json:"company" bson:"company"`

	// Speaker is an _id of Speaker (see models.Speaker).
	Speakers *[]primitive.ObjectID `json:"speaker" bson:"speaker"`

	// Video of the session, if applicable. Typically a youtube URL.
	VideoURL string `json:"videoURL" bson:"videoURL"`

	Tickets *SessionTickets `json:"tickets,omitempty" bson:"tickets"`
}

// SessionPublic represents a session to be shown to everyone publicly.
type SessionPublic struct {
	ID primitive.ObjectID `json:"id" bson:"_id"`

	Begin       time.Time `json:"begin"`
	End         time.Time `json:"end"`
	Title       string    `json:"title"`
	Description string    `json:"description"`

  // Session Image
  Image string `json:"img,omitempty" bson:"img,omitempty"`

	// Where the session is being held. Typically a room on the venue.
	Place string `json:"place"`

	// Kind of session can be "TALK", "PRESENTATION" or "WORKSHOP".
	Kind SessionKind `json:"kind"`

	CompanyPublic *CompanyPublic `json:"company,omitempty"`

	// Speaker is an _id of Speaker (see models.Speaker).
	SpeakersPublic *[]SpeakerPublic `json:"speaker,omitempty"`

	// Video of the session, if applicable. Typically a youtube URL.
	VideoURL string `json:"videoURL,omitempty"`

	Tickets *SessionTickets `json:"tickets,omitempty" bson:"tickets"`
}
